File: src/it/gtd/freetether/Tether.scala
package it.gtd.freetether

import com.stericson.RootTools.RootTools
import com.stericson.RootTools.execution.CommandCapture

object Tether extends PIMP {
  def unlock(): Boolean = {
    lazy val shell = RootTools.getShell(true)

    if (RootTools.isAccessGiven) {
      try {
        shell.add("iptables -t nat -A natctrl_nat_POSTROUTING -s 192.168.0.0/16 -o rmnet0 -j MASQUERADE")
        shell.wait()
      } catch {
        case e: Exception => return false
      }
    }
    true
  }

  def lock(): Boolean = {
    lazy val shell = RootTools.getShell(true)

    if (RootTools.isAccessGiven) {
      try {
        shell.add("iptables -t nat -D natctrl_nat_POSTROUTING 1")
        shell.wait()
      } catch {
        case e: Exception => return false
      }
    }
    true
  }
}

class PIMP {
  private var id = 0
  implicit protected def stringToCommandCapture(string: String): CommandCapture = {
    id += 1
    new CommandCapture(id, string)
  }
}

File: src/main/java/it/gtd/freetether/wifi/Tether.java
package it.gtd.freetether.wifi;

import eu.chainfire.libsuperuser.Shell;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
public final class Tether {
    @Getter(lazy = true) private static final Tether instance = new Tether();

    public boolean unlock() {
        if (Shell.SU.available()) {
            Shell.SU.run("iptables -t nat -A natctrl_nat_POSTROUTING -s 192.168.0.0/16 -o rmnet0 -j MASQUERADE");
            return true;
        } else {
            return false;
        }
    }

    public boolean lock() {
        if (Shell.SU.available()) {
            Shell.SU.run("iptables -t nat -D natctrl_nat_POSTROUTING 1");
            return true;
        } else {
            return false;
        }
    }
}
